crates/p2p/src/behaviour.rs

```diff
@@ -468,6 +480,59 @@ impl Behaviour {
         Ok(())
     }
 
+    /// Evict an outbound peer if the maximum number of outbound peers has been reached.
+    fn evict_outbound_peer(&mut self) -> Result<(), ConnectionDenied> {
+        let mut candidates: Vec<_> = self.outbound_peers().collect();
+        if candidates.len() < self.cfg.max_outbound_peers {
+            return Ok(());
+        }
+
+        // Only peers which are flagged as not useful are considered for eviction.
+        candidates.retain(|(_, peer)| !peer.useful);
+
+        // The peer to be evicted is the one with the highest SHA3(eviction_secret || peer_id)
+        // value. This is deterministic but unpredictable by any outside observer.
+        candidates.sort_by_key(|(peer_id, _)| {
+            use sha3::{Digest, Sha3_256};
+            let mut hasher = Sha3_256::default();
+            hasher.update(self.eviction_secret);
+            hasher.update(peer_id.to_bytes());
+            hasher.finalize()
+        });
+        let (peer_id, _) = candidates.pop().ok_or_else(|| {
+            tracing::debug!("Outbound peer limit reached, but no peers could be evicted");
+            ConnectionDenied::new(anyhow!(
+                "outbound peer limit reached and no peers could be evicted"
+            ))
+        })?;
```

Comment:
Can also be done using the `let Some(_) = .. else { .. };` syntax which saves a closure and highlights the control flow 🤷